src/test/mod.rs

```diff
@@ -44,3 +36,45 @@ impl TestDatabase {
         fakes::FakeRelease::new(self)
     }
 }
+
+pub(crate) fn with_frontend(db: &TestDatabase, f: impl FnOnce(&TestFrontend) -> Result<(), Error>) {
+    let frontend = TestFrontend::new(db);
+    let result = f(&frontend);
+    frontend.server.leak();
+    report_error(|| result)
+}
+
+pub(crate) struct TestFrontend {
+    server: Server,
+    client: Client,
+}
+
+impl TestFrontend {
+    fn new(db: &TestDatabase) -> Self {
+        Self {
+            server: Server::start_test(db.conn.clone()),
+            client: Client::new(),
+        }
+    }
+
+    fn build_request(&self, method: Method, url: &str) -> RequestBuilder {
+        self.client.request(method, &format!("http://{}{}", self.server.addr(), url))
+    }
+
+    pub(crate) fn get(&self, url: &str) -> RequestBuilder {
+        self.build_request(Method::GET, url)
+    }
+}
+
+fn report_error(f: impl FnOnce() -> Result<(), Error>) {
```

Comment:
It might be simpler to have this take a `Result<(), Error>` instead of passing it a closure directly. It doesn't look like there's any setup or teardown so it shouldn't affect the semantics.

Reply:
Whoops, this function got removed in a later commit.